src/Domain/Order/Handler/CheckoutHandler.php

```diff
@@ -23,56 +23,59 @@ public function __construct(
     {
     }
 
-    private function getLastPayment(OrderInterface $order): ?PaymentInterface
-    {
-        if ($payment = $order->getLastPayment(PaymentInterface::STATE_CART)) {
-
-            return $payment;
-        }
-
-        return null;
-    }
-
     public function __invoke(Checkout $command)
     {
         $order = $command->getOrder();
         $stripeToken = $command->getStripeToken();
 
-        $payment = $this->getLastPayment($order);
-
-        $isFreeOrder = null === $payment && $order->isFree();
-        $isCashOnDelivery = null !== $payment && $payment->isCashOnDelivery();
-        $isMercadopagoPaymentApproved = null !== $payment && 'approved' === $payment->getMercadopagoPaymentStatus();
-
-        if ($isFreeOrder || $isCashOnDelivery || $isMercadopagoPaymentApproved) {
-            $this->orderNumberAssigner->assignNumber($order);
-            $this->eventRecorder->record(new Event\CheckoutSucceeded($order, $payment));
+        // Assign a number to the order in any case
+        $this->orderNumberAssigner->assignNumber($order);
 
+        // Bail early if order is free
+        if ($order->isFree()) {
+            $this->eventRecorder->record(new Event\CheckoutSucceeded($order));
             return;
         }
 
-        // TODO Check if $payment !== null
+        $payments = $order->getPayments()->filter(
+            fn (PaymentInterface $payment): bool => $payment->getState() === PaymentInterface::STATE_CART);
 
-        $data = $command->getData();
+        // TODO Make sure card payment is always *BEFORE* Edenred payment
```

Comment:
it is just sorting the payments array right?

Reply:
I added this comment because it was [hardcoded like this](https://github.com/coopcycle/coopcycle-web/blob/f74674c9c7706cac381f29d8fc471d7420a45d4f/src/Payment/Gateway.php#L52-L74) before. 

I don't remember exactly why, probably because it's easier to re-make a card authorization than an Edenred authorization (Edenred authorizations **MUST** be [cancelled when the order is refused/cancelled](https://github.com/coopcycle/coopcycle-web/blob/f74674c9c7706cac381f29d8fc471d7420a45d4f/src/Domain/Order/Reactor/CancelEdenredTransaction.php), to make sure we release the money).